src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -39,79 +40,76 @@ import scala.annotation.tailrec
 
-  // invariant: if "_tasksLocal.get ne null" then we are inside BatchingRunnable.run; if it is null, we are outside
-  private val _tasksLocal = new ThreadLocal[List[Runnable]]()
+  private[this] final class Batch(size: Int) extends ArrayDeque[Runnable](size) with Runnable with BlockContext with (BlockContext => Unit) {
+    private[this] final var parentBlockContext: BlockContext = _
 
-  private class Batch(val initial: List[Runnable]) extends Runnable with BlockContext {
-    private var parentBlockContext: BlockContext = _
-    // this method runs in the delegate ExecutionContext's thread
-    override def run(): Unit = {
-      require(_tasksLocal.get eq null)
+    def this(r: Runnable) = {
+      this(4)
+      addLast(r)
+    }
 
-      val prevBlockContext = BlockContext.current
-      BlockContext.withBlockContext(this) {
-        try {
-          parentBlockContext = prevBlockContext
+    final def executor: BatchingExecutor = BatchingExecutor.this
 
-          @tailrec def processBatch(batch: List[Runnable]): Unit = batch match {
-            case Nil => ()
-            case head :: tail =>
-              _tasksLocal set tail
-              try {
-                head.run()
-              } catch {
-                case t: Throwable =>
-                  // if one task throws, move the
-                  // remaining tasks to another thread
-                  // so we can throw the exception
-                  // up to the invoking executor
-                  val remaining = _tasksLocal.get
-                  _tasksLocal set Nil
-                  unbatchedExecute(new Batch(remaining)) //TODO what if this submission fails?
-                  throw t // rethrow
-              }
-              processBatch(_tasksLocal.get) // since head.run() can add entries, always do _tasksLocal.get here
-          }
+    // this method runs in the delegate ExecutionContext's thread
+    override final def run(): Unit = BlockContext.usingBlockContext(this)(this)
+
+    override final def apply(prevBlockContext: BlockContext): Unit = {
+      //This invariant needs to hold: require(_tasksLocal.get eq null)
+      parentBlockContext = prevBlockContext
+      try {
+        _tasksLocal.set(this)
+        runAll()
+        _tasksLocal.remove() // Will be cleared in the throwing-case by runAll()
+      } finally {
+        parentBlockContext = null
+      }
+    }
 
-          processBatch(initial)
-        } finally {
-          _tasksLocal.remove()
-          parentBlockContext = null
-        }
+    @tailrec private[this] final def runAll(): Unit = {
+      val next = pollLast()
+      if (next ne null) {
+        try next.run() catch {
+          case t: Throwable =>
+            parentBlockContext = null // Need to reset this before re-submitting it
+            _tasksLocal.remove() // If unbatchedExecute runs synchronously
+            unbatchedExecute(this) //TODO what if this submission fails?
```

Comment:
could potentially

```scala
try unbatchedExecute(this)
catch {
  case inner: Throwable if NonFatal(t) && !NonFatal(inner) =>
    inner.addSuppressed(t)
    throw inner
  case inner: Throwable => t.addSuppressed(inner)
}
```

Reply:
Yeah, that doesn't play well with cached exception instances though. Thoughts?

Reply:
A few possibilities I've come up with:

- if you're not concerned about `unbatchedExecute` throwing a cached exception instance, you could always `addSuppressed` to that one, even if it would in general make more sense to rethrow the original one
- construct a new exception with one as the cause and the other suppressed. I'm not super happy with this, but it certainly solves the cached exception problem

Reply:
@NthPortal How about this:

~~~scala
    private[this] final def handleRunFailure(cause: Throwable): Unit =
      if (NonFatal(t) || t.isInstanceOf[InterruptedException]) {
        try unbatchedExecute(this) catch {
          case inner: Throwable =>
            if (NonFatal(inner)) {
              val e = new ExecutionException("Non-fatal error occurred and resumission failed, see suppressed exception.", t)
              e.addSuppressed(inner)
              throw e
            } else throw inner // inner exception is fatal, throw it
        }
      } else throw t // throw fatals
~~~

Reply:
OTOH it needs some extra consideration w.r.t. Interruptions. Thinking…

Reply:
I believe this will work @NthPortal:

~~~scala
    private[this] final def handleRunFailure(cause: Throwable): Nothing =
      if (NonFatal(cause) || cause.isInstanceOf[InterruptedException]) {
        try unbatchedExecute(this) catch {
          case inner: Throwable =>
            if (NonFatal(inner)) {
              val e = new ExecutionException("Non-fatal error occurred and resubmission failed, see suppressed exception.", cause)
              e.addSuppressed(inner)
              throw e
            } else throw inner
        }
        throw cause
      } else throw cause
~~~

Reply:
Rationale: original exception will be rethrown if it is fatal or if unbatchedExecute is successful.
If unbatchedExecute throws a Fatal (includes InterruptedException) then just rethrow that, but if  not, then both the original exception (cause) and the inner exception is wrappen in a new exception and thrown